src/Model/Payload.php

```diff
@@ -113,4 +110,37 @@ public function getCustomData(): array
     {
         return $this->customData;
     }
+
+    /**
+     * Guard the custom data upon instantiation.
+     *
+     * @param array $data
+     */
+    private function guardCustomData(array $data)
+    {
+        foreach ($data as $key => $value) {
+            $this->validateCustomDataValue($value);
+            $this->validateCustomDataKey($key);
+        }
+    }
+
+    private function validateCustomDataValue($value)
```

Comment:
Why we have the `validateCustomDataValue` and `validateCustomDataKey` functions?

We can join this methods to one method and check the key and value in new method:

```php
private function validateCustomData(array $customData): void
{
    foreach ($customData as $key => $value) {
        // Validate key
        // Validate value

        if (is_array($value)) {
            $this->validateCustomData($value);
        }
    }
}
```

Note: after these changes, we can check the hierarchical arrays with use loops.

Implementation for `withCustomData($key, $value)` function:

```php
public function withCustomData($key, $value)
{
    $this->validateCustomData([$key => $value]);
}
```